tree/dataframe/inc/ROOT/RDF/RAction.hxx

```diff
@@ -139,22 +139,28 @@ public:
       // Action nodes do not need to go through CreateFilterNode: they are never common nodes between multiple branches
       auto thisNode = std::make_shared<RDFGraphDrawing::GraphNode>(fHelper.GetActionName());
 
-      auto evaluatedNode = thisNode;
-      for (auto &column : GetDefines().GetColumns()) {
-         /* Each column that this node has but the previous hadn't has been defined in between,
-          * so it has to be built and appended. */
-         if (RDFInternal::IsInternalColumn(column.first))
-            continue;
-         if (std::find(prevColumns.begin(), prevColumns.end(), column.first) == prevColumns.end()) {
-            auto defineNode = RDFGraphDrawing::CreateDefineNode(column.first, column.second.get());
-            evaluatedNode->SetPrevNode(defineNode);
-            evaluatedNode = defineNode;
-         }
+      // Add the Defines that have been added between this node and the previous to the graph
+      auto upmostNode = thisNode;
+      const auto &defineNames = GetDefines().GetNames();
+      const auto &defineMap = GetDefines().GetColumns();
+      for (auto i = int(defineNames.size()) - 1; i >= 0; --i) {  // walk backwards through the names of defined columns
```

Comment:
What's the complexity of this impl? Is it a N^2-like operation in every node of the graph?

Reply:
no no, this checks just the defines that have been added between the previous non-Define node in the graph and this one -- due to the `break` 6 lines below.

And in general the chain of recursive `GetGraph` calls is cut short whenever we reach a node that was already visited.

Reply:
Alright, thanks! :+1: